File: tl/src/main/java/com/github/badoualy/telegram/tl/api/requests/TLRequestChannelsCreateChannel.java

package com.github.badoualy.telegram.tl.api.requests;

import com.github.badoualy.telegram.tl.TLContext;
import com.github.badoualy.telegram.tl.core.TLMethod;
import com.github.badoualy.telegram.tl.core.TLObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLObject;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLBool;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLString;


public class TLRequestChannelsCreateChannel extends TLMethod<com.github.badoualy.telegram.tl.api.TLAbsUpdates> {

    public static final int CLASS_ID = 0xf4893d7f;

    public int getClassId() {
        return CLASS_ID;
    }


    public TLRequestChannelsCreateChannel(        int _flags,         boolean _broadcast,         boolean _megagroup,         String _title,         String _about) {
        this.flags = _flags;
        this.broadcast = _broadcast;
        this.megagroup = _megagroup;
        this.title = _title;
        this.about = _about;

    }



    public com.github.badoualy.telegram.tl.api.TLAbsUpdates deserializeResponse(InputStream stream, TLContext context) throws IOException {

        TLObject res = readTLObject(stream, context);
        if (res == null) {
            throw new IOException("Unable to parse response");
        }
        if (res instanceof com.github.badoualy.telegram.tl.api.TLAbsUpdates) {
            return (com.github.badoualy.telegram.tl.api.TLAbsUpdates)res;
        }
        else {
            throw new IOException("Incorrect response type. Expected com.github.badoualy.telegram.tl.api.TLAbsUpdates, got: " + res.getClass().getCanonicalName());
        }

    }
        


    protected int flags;

    protected boolean broadcast;

    protected boolean megagroup;

    protected String title;

    protected String about;


    public int getFlags() {
        return flags;
    }

    public void setFlags(int value) {
        this.flags = value;
    }

    public boolean getBroadcast() {
        return broadcast;
    }

    public void setBroadcast(boolean value) {
        this.broadcast = value;
    }

    public boolean getMegagroup() {
        return megagroup;
    }

    public void setMegagroup(boolean value) {
        this.megagroup = value;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String value) {
        this.title = value;
    }

    public String getAbout() {
        return about;
    }

    public void setAbout(String value) {
        this.about = value;
    }


    @Override
    public void serializeBody(OutputStream stream) throws IOException {

        flags = broadcast ? (flags | 1) : (flags &~ 1);
        flags = megagroup ? (flags | 2) : (flags &~ 2);
        writeInt(this.flags, stream);
        if ((this.flags & 1) != 0)
            writeTLBool(this.broadcast, stream);
        if ((this.flags & 2) != 0)
            writeTLBool(this.megagroup, stream);
        writeTLString(this.title, stream);
        writeTLString(this.about, stream);
    }


    @Override
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {

        this.flags = readInt(stream);
        this.broadcast = (this.flags & 1) != 0;
        this.megagroup = (this.flags & 2) != 0;
        this.title = readTLString(stream);
        this.about = readTLString(stream);
    }



    @Override
    public String toString() {
        return "channels.createChannel#f4893d7f";
    }

}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/request/TLRequestChannelsCreateChannel.kt
package com.github.badoualy.telegram.tl.api.request

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.api.TLAbsUpdates
import com.github.badoualy.telegram.tl.core.TLMethod
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.jvm.Throws
import kotlin.jvm.Transient

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLRequestChannelsCreateChannel() : TLMethod<TLAbsUpdates>() {
    @Transient
    var broadcast: Boolean = false

    @Transient
    var megagroup: Boolean = false

    var title: String = ""

    var about: String = ""

    private val _constructor: String = "channels.createChannel#f4893d7f"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            broadcast: Boolean,
            megagroup: Boolean,
            title: String,
            about: String
    ) : this() {
        this.broadcast = broadcast
        this.megagroup = megagroup
        this.title = title
        this.about = about
    }

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(broadcast, 1)
        updateFlags(megagroup, 2)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        writeString(title)
        writeString(about)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        broadcast = isMask(1)
        megagroup = isMask(2)
        title = readString()
        about = readString()
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += computeTLStringSerializedSize(title)
        size += computeTLStringSerializedSize(about)
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLRequestChannelsCreateChannel) return false
        if (other === this) return true

        return _flags == other._flags
                && broadcast == other.broadcast
                && megagroup == other.megagroup
                && title == other.title
                && about == other.about
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0xf4893d7f.toInt()
    }
}
